### DISCUSSION TITLE
求一份关于表格的详细说明

### DISCUSSION THREAD

**Discussion by dreistein543** [nju-lug/NJUThesis#97] [Q&A]
目前`njuthesis-v0.15.0.pdf`文档中关于表格的描述过于简单，实际使用中，一般面临如下种情况：

1. 表格大小合适
2. 表格过长 （尝试使用过 longtable 包，但似乎会与 booktabs 设置冲突）
3. 表格过宽 （尝试过 sidewaystable ，但是似乎在识别奇偶页上不够智能，现在用 landscape 配合 {\hsize} 比较好解决）
4. 表格又长又宽 （见如下的 widetable 结果，横向后依然可能过长）

就我目前遇到的情况来看，没有办法简单解决，因为即使大小合适的表格，也可能存在一些问题（在单独页面不顶格而是上下居中显示），希望能针对该情况出一份比较详细的说明。（曾经尝试过 resizebox 但是玩不转 TAT）
使用配置：

- M1 mactex 2021 本地编译
- texstudio latexmk -xelatex 方式编译

代码如下：

```latex
\usepackage{longtable} % 绘制长表格用
\usepackage{booktabs} % 表格样式
\usepackage{rotating} % 表格横向
\usepackage{dcolumn}  % 小数点对齐
\usepackage{lscape} % 页面横向
\usepackage{lipsum} % 简单生成大段废话观察结果

\begin{document}

\input{mytable1}

\lipsum{3}

\begin{landscape}
	\input{widetable}
\end{landscape}

\lipsum{5}

\input{longtable}

\end{document}
```

使用的表格tex文件附上（需修改后缀）
[longtable.txt](https://github.com/nju-lug/NJUThesis/files/8230769/longtable.txt)
[widetable.txt](https://github.com/nju-lug/NJUThesis/files/8230770/widetable.txt)
[mytable1.txt](https://github.com/nju-lug/NJUThesis/files/8230771/mytable1.txt)

**Comment by atxy-blip** [comment]
上周比较忙，回复迟了点。我的解决方法是换用功能更全的 `tabularray` 包：

```TeX
\documentclass{njuthesis}
\usepackage{tabularray,zhlipsum}
\UseTblrLibrary{booktabs}

\begin{document}
\mainmatter
\chapter{表格}
\section{长表}
\zhlipsum[1]
\input{longtable}
\end{document}
```

更新后的 `longtable.tex` 见下方，可以用同一段内容重复来扩充长度：

> 除了原文中带 $\times$ 的两行都不会超出目前的页边距。

```TeX

\begin{longtblr}[
    caption = {regResults},
    label = {tab:long},
    remark{Note} = {Exponentiated coefficients; $t$ statistics in parentheses},
    note{a} = {\(p<0.05\)},
    note{b} = {\(p<0.01\)},
    note{c} = {\(p<0.001\)},
    note{d} = {\(p<0.1\)}
    ]{colspec = {ccccccccc}, rows = {font= \small}, rowsep = 0pt}
    \toprule
                    & \SetCell[c=4]{c} main & & &
                    & \SetCell[c=4]{c} interaction                                       \\
    \cmidrule[lr]{2-5} \cmidrule[lr]{6-9}
                    & \SetCell[c=2]{c} (1)          & & \SetCell[c=2]{c} (2) &
                    & \SetCell[c=2]{c} (3)          & & \SetCell[c=2]{c}(4)              \\
    \midrule
    OveSeaEdu       &    1.193\TblrNote{c} &  (3.583) &    1.186\TblrNote{c}  &  (3.459)
                    &    1.142\TblrNote{a} &  (2.429) &    1.186\TblrNote{c}  &  (3.459) \\
    FemalePost80    &    1.218             &  (1.600) &    1.104              &  (0.779)
                    &    1.094             &  (0.713) &    1.010              &  (0.046) \\
    Lev             &    1.859\TblrNote{c} &  (4.049) &    1.884\TblrNote{c}  &  (4.131)
                    &    1.892\TblrNote{c} &  (4.153) &    1.880\TblrNote{c}  &  (4.114) \\
    Cash            &    0.969\TblrNote{d} & (-1.869) &    0.970\TblrNote{d}  & (-1.812)
                    &    0.970\TblrNote{d} & (-1.786) &    0.970\TblrNote{d}  & (-1.825) \\
    \bottomrule
\end{longtblr}
```

编译出来就是图中这个效果。我斗胆提一点建议：

1. 原文中带括号的数据另起一行，是否会增加不必要的长度，以及造成阅读上的困难？
2. *Industry* 是否可以作为表格的一行存在，考虑到其中的条目数量，是否可以另起一张表？
3. 这种提供原始数据的长表格是不是放在附录中更合适？

![image](https://user-images.githubusercontent.com/63113546/158048630-72d9c1bd-e11b-4a5a-bb63-c062ff7c49bb.png)

**Reply by dreistein543** [reply]
感谢大佬的回复，事实上，tex表格的输出是由所用的统计分析软件输出的，格式也是按照学院以及大部分期刊的格式规范来的（例如括号行另起一行），industry内容过多只是为了塑造一个等宽的“长表格”（通常与year一样只需要一个yes即可），但是事实上，只要控制变量足够多（往往也是非常需要的），正文分析表格至少有一张会超越A4纸长度，所以通过修改表格内容以及更改显示方式的方法避免表格溢出属于不能保证的方法（最多只能接受全局修改变量名压缩页宽）。
另外，我注意到跨页的部分标记（continued on next page）以及续表标记（continued）都是英文，且您给出的语法部分并无可以自行定义的位置，不知道能否指点一下如何将其修改为中文内容🙏。即可以没有“剩余部分见下页”但是续表最好标题是“续表xxx”，实现起来复杂的话可以接受没有这两块的标记

**Reply by atxy-blip** [reply]
`tabularray` 宏包手册（2022A 版）39 页就写了怎么定义这个文字，在导言区加两句命令就行：

```TeX
\DefTblrTemplate{contfoot-text}{default}{接下一页}
\DefTblrTemplate{conthead-text}{default}{（续）}
```

> 什么，你找不到手册？终端里输入 `texdoc tabularray` 或者[在  CTAN 找](http://mirrors.ctan.org/macros/latex/contrib/tabularray/tabularray.pdf)

这个包提供了各种表格样式的一揽子解决方案。稍微熟悉一下新的接口，用起来还是很舒服的。

**Reply by dreistein543** [reply]
在使用上还遇到一些小问题：
1. 第一个是关于列对齐的问题

根据`tabularry`包的介绍，它只提供了一种代号为 `Q` 的对齐方式（column type，然而并不明白Q对齐是什么意思），默认支持 lrc（左右中对齐）
但是经我尝试，它不能识别 `dcolumn` 中的 `D{.}{.}{-1}` 小数点对齐，也不能识别 `siunitx` 中 `S[table-format=#.#]`的小数点对齐
据包使用说明介绍，似乎需要自行定义其他的column type（如下图展示了包说明中所有定义案例），然而并看不懂如何去定义小数点对齐TAT（而且包说明文件也没有使用数值举例，尝试搜索 decimal align 也没有结果）。还请大佬指点一下
<img width="862" alt="截屏2022-11-08 23 54 10" src="https://user-images.githubusercontent.com/30745407/200612727-47ac9f87-d108-48df-a484-e9ba557ed0ce.png">

2. 使用 `tabularray` 包遇到了奇怪的bug

macOS13  本地编译
我在本地尝试编译的时候，第二个大列（边际效应）不会显示出来。不清楚为什么(没有报错或者警告)
在 overleaf 中尝试了一下发现也没办法展示

```latex
\def\sym#1{\ifmmode^{#1}\else\(^{#1}\)\fi} % 定义 \sym{**} 为上标星星
\begin{longtblr}[
    caption = {随便写点},
    label = {tbl:main},
    remark{注} = {随便写点}
    ]{colspec = {lccccc}, rows = {font= \small}, rowsep = 0pt}
    \toprule
                &         &\SetCell[c=2]{c}{Probit模型系数}   &\SetCell[c=2]{c}{边际效应}         \\\cmidrule[lr]{3-4}\cmidrule[lr]{5-6}
                &预测方向 &\SetCell[c=1]{c}{(1)}         &\SetCell[c=1]{c}{(2)}         &\SetCell[c=1]{c}{(3)}         &\SetCell[c=1]{c}{(4)}         \\
\midrule
\hspace{0.1cm} MainVar&    +    &                  &    0.000\sym{***}&                  &    0.000\sym{***}\\
                &         &                  &  (0.000)         &                  &  (0.000)         \\
\emph{变量分组} &         &                  &                  &                  &                  \\
\hspace{0.1cm} ModVar&         &    0.000\sym{**} &    0.000\sym{***}&    0.000\sym{**} &    0.000\sym{***}\\
                &         &  (0.000)         &  (0.000)         &  (0.000)         &  (0.000)         \\
    \bottomrule
\end{longtblr}
```
<img width="648" alt="截屏2022-11-08 23 43 34" src="https://user-images.githubusercontent.com/30745407/200610104-a55ac092-3b6b-4cea-878f-e99a734f8de7.png">

**Reply by dreistein543** [reply]
找到了，只要设定 `\UseTblrLibrary{siunitx}` 就可以继续使用 `S` 作为对齐方式，但是出现了其他问题。
1. 首先是，如果表格中出现了字符信息，就直接报错，即使我将字符使用 `\SetCell{c=1}{c}{字符}` 指定字符的对齐方式为居中也还是报错
2. 对齐方式无法精细控制，在常规的表格中，指定 `S` 作为小数点居中时，可以详细设定 `S[table-format=2.3]` 来设定小数点前后的字符宽度，这里无法设定，至少是无法简单设定（支持通过`colsep={lc*{4}S}`来设置4个小数点对齐，也支持重复的`colsep={lcS[table-format=2.3]S[table-format=2.3]S[table-format=2.3]S[table-format=2.3]}`,但就是不支持`colsep={lr*{4}S[table-format=2.3]}`）
3. `\SetCell{c=1}{数字}`来指定某一单元格的对齐方式为居中无效（虽然不报错，但依旧是将其作为小数点对齐的整数部分）

```LaTeX
\usepackage{tabularrary}
\UseTblrLibrary{siunitx}

\def\sym#1{\ifmmode^{#1}\else\(^{#1}\)\fi} % 定义 \sym{**} 为上标星星
\begin{longtblr}[
    caption = {随便写点},
    label = {tbl:main},
    remark{注} = {随便写点}
    ]{colspec = {lc*{4}S}, rows = {font= \small}, rowsep = 0pt}
    \toprule
                &         &\SetCell[c=2]{c}{Probit模型系数}   &\SetCell[c=2]{c}{边际效应}         \\\cmidrule[lr]{3-4}\cmidrule[lr]{5-6}
                &预测方向 &\SetCell[c=1]{c}{(1)}         &\SetCell[c=1]{c}{(2)}         &\SetCell[c=1]{c}{(3)}         &\SetCell[c=1]{c}{(4)}         \\
\midrule
\hspace{0.1cm} MainVar&    +    &                  &    0.000\sym{***}&                  &    0.000\sym{***}\\
                &         &                  &  (0.000)         &                  &  (0.000)         \\
\emph{变量分组} &         &                  &                  &                  &                  \\
\hspace{0.1cm} ModVar&         &    0.000\sym{**} &    0.000\sym{***}&    0.000\sym{**} &    0.000\sym{***}\\
                &         &  (0.000)         &  (0.000)         &  (0.000)         &  (0.000)         \\
% Year            &         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         \\
% Industry        &         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         &\SetCell[c=1]{c}{Yes}         \\
\midrule
样本量        &         &\SetCell[c=1]{c}{10000}    &\SetCell[c=1]{c}{10000}         &\SetCell[c=1]{c}{10000}         &\SetCell[c=1]{c}{10000}         \\
  \bottomrule
\end{longtblr}
```
可以看到，模型序号不会居中，最下方的样本量整数不居中，对行业和年份的Yes如果取消注释就会报错
![截屏2022-11-09 01 21 16](https://user-images.githubusercontent.com/30745407/200632706-3b0df230-52b0-4b4b-9a7a-8f175122287a.png)

**Reply by dreistein543** [reply]
还是准备放弃 `tabularray` 包的使用了

> 我在本地尝试编译的时候，第二个大列（边际效应）不会显示出来。不清楚为什么(没有报错或者警告)

这是因为 `tabularray` 的 \SetCell 指令强制要求原本 \multicolumn{} 中经由合并可以省略的列，在使用 \setcell 时不可以省略，
也就意味着，必须在 \SetCell[c=2]{} 后面加上 `& &` 认为指定一个空行。这给我利用其他软件直接导出 tex文件带来极大的麻烦（本来已经需要花力气去调整默认的 \multicolumn 为 \setcell）

不仅如此，也不知道为什么 colspec={} 指令似乎在本地必须写成一行才能运行（明明作者的github案例都可以写成多行）

除此以外，表格中除了表头和第一列，完全不能出现字符，明明通过制定对齐方式 S[table-format=2.3,c] 可以指定非数字为居中对齐，就是疯狂报错。我已经不敢再赌了

**Reply by atxy-blip** [reply]
这些关于 `tabularray` 宏包的问题或许应该提到 <https://github.com/lvjr/tabularray/discussions>，我作为模板作者显然没法对宏包的运行机制负责……

工具是死的，人是活的，这个不行就换个用着顺手的。
